legate/core/legion.py

```diff
@@ -2696,7 +2862,8 @@ def __init__(
         legion.legion_memory_query_destroy(query)
         legion.legion_machine_destroy(machine)
         for (sub_region, buf) in shard_local_data.items():
-            assert sub_region.parent.parent is parent
+            if sub_region.parent is not None:
+                assert sub_region.parent.parent is parent
```

Comment:
I assume assertions were used when it's an internal error. Anything in the legion module is only supposed to be used by legate developers, not users.

Reply:
Projects will often define their own `InternalError` or similar for this purpose (e.g. [here](https://www.tensorflow.org/api_docs/python/tf/errors/InternalError), [here](https://dev.mysql.com/doc/connector-python/en/connector-python-api-errors-internalerror.html), or [here](https://www.adamsmith.haus/python/docs/keras.losses.K.ctc.gen_ctc_ops._errors.InternalError)) This allows for the possibility of attaching more detailed diagnostic information or state information to the error (which can be used by legate developers, or reported by users making issues, improving the quality of bug reports)

Reply:
> sub_region.parent can technically None. Perhaps there is better solution, though.

It would be nice if we could actually enforce this at the typing level, say making a subtype `SubRegion` which does actually the same thing as `Region` but always has a non-null `parent`. Then the caller needs to do a casting to make sure the argument has a right type.

> I assume assertions were used when it's an internal error. Anything in the legion module is only supposed to be used by legate developers, not users.

@manopapad unless this class is consumed completely within Legate core, I suggest we instead raise an exception with a more meaningful message so that the library developer can learn the right usage of this API.

Reply:
:+1: to both those suggestions but I given the scope of changes already in this PR I would suggest and issue for now and visiting after `legion.py` is split up

Reply:
sounds good to me.